File: 42. Trapping Rain Water.go
func trap(height []int) int {
	answer := 0
	stack := []int{}
	for i := range height {
		for len(stack) > 0 && height[i] >= height[stack[len(stack)-1]] {
			h := height[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]
			if len(stack) > 0 {
				answer += (i - stack[len(stack)-1] - 1) * (min(height[i], height[stack[len(stack)-1]]) - h)
			}
		}
		stack = append(stack, i)
	}
	return answer
}

File: 42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        answer = 0
        stack = []
        stack.append(0)
        for i in range(len(height)):
            while stack and height[i] >= height[stack[-1]]:
                h = height[stack.pop()]
                if stack:
                    answer += (i-stack[-1]-1)*(min(height[i], height[stack[-1]])-h)
            stack.append(i)
        return answer